[quant][core][gpu][improvement] Added support for padding quantized cudnn conv2d operator

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76184

Summary:
cudnn v8.4.0 expects input channels for conv2d to be a multiple of 4. If
it is not, we need to explicitly pad it to a multiple of 4 ourselves as
cudnn does not currently support padding intriniscally.
The padding implemented here is limited to groups=1; however, this
should be a straightforward adaption to groups > 1 since we're only
padding a single dimension.

When cudnn enables support for padding, we can remove the padding on our
end.

Test plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

Differential Revision: [D35822988](https://our.internmc.facebook.com/intern/diff/D35822988)

dzdang committed:
[quant][core][gpu][improvement] ] Added support for padding quantized cudnn conv2d operator

Summary:
cudnn v8.4.0 expects input channels for conv2d to be a multiple of 4. If
it is not, we need to explicitly pad it to a multiple of 4 ourselves as
cudnn does not currently support padding intriniscally.
The padding implemented here is limited to groups=1; however, this
should be a straightforward adaption to groups > 1 since we're only
padding a single dimension.

When cudnn enables support for padding, we can remove the padding on our
end.

Test plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][improvement] Added support for padding quantized cudnn conv2d operator"


Summary:
cudnn v8.4.0 expects input channels for conv2d to be a multiple of 4. If
it is not, we need to explicitly pad it to a multiple of 4 ourselves as
cudnn does not currently support padding intriniscally.
The padding implemented here is limited to groups=1; however, this
should be a straightforward adaption to groups > 1 since we're only
padding a single dimension.

When cudnn enables support for padding, we can remove the padding on our
end.

Test plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35822988).

dzdang committed:
Update on "[quant][core][gpu][improvement] Added support for padding quantized cudnn conv2d operator"


Summary:
cudnn v8.4.0 expects input channels for conv2d to be a multiple of 4. If
it is not, we need to explicitly pad it to a multiple of 4 ourselves as
cudnn does not currently support padding intriniscally.
The padding implemented here is limited to groups=1; however, this
should be a straightforward adaption to groups > 1 since we're only
padding a single dimension.

When cudnn enables support for padding, we can remove the padding on our
end.

Test plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

Differential Revision: [D35822988](https://our.internmc.facebook.com/intern/diff/D35822988)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35822988).

dzdang committed:
Update on "[quant][core][gpu][improvement] Added support for padding quantized cudnn conv2d operator"


Summary:
cudnn v8.4.0 expects input channels for conv2d to be a multiple of 4. If
it is not, we need to explicitly pad it to a multiple of 4 ourselves as
cudnn does not currently support padding intriniscally.
The padding implemented here is limited to groups=1; however, this
should be a straightforward adaption to groups > 1 since we're only
padding a single dimension.

When cudnn enables support for padding, we can remove the padding on our
end.

Test plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

Differential Revision: [D35822988](https://our.internmc.facebook.com/intern/diff/D35822988)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][improvement] Added support for padding quantized cudnn conv2d operator"


Summary:
cudnn v8.4.0 expects input channels for conv2d to be a multiple of 4. If
it is not, we need to explicitly pad it to a multiple of 4 ourselves as
cudnn does not currently support padding intriniscally.
The padding implemented here is limited to groups=1; however, this
should be a straightforward adaption to groups > 1 since we're only
padding a single dimension.

When cudnn enables support for padding, we can remove the padding on our
end.

Test plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

Differential Revision: [D35822988](https://our.internmc.facebook.com/intern/diff/D35822988)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35822988).

dzdang committed:
Update on "[quant][core][gpu][improvement] Added support for padding quantized cudnn conv2d operator"


Summary:
cudnn v8.4.0 expects input channels for conv2d to be a multiple of 4. If
it is not, we need to explicitly pad it to a multiple of 4 ourselves as
cudnn does not currently support padding intriniscally.
The padding implemented here is limited to groups=1; however, this
should be a straightforward adaption to groups > 1 since we're only
padding a single dimension.

When cudnn enables support for padding, we can remove the padding on our
end.

Test plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

Differential Revision: [D35822988](https://our.internmc.facebook.com/intern/diff/D35822988)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35822988).

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/conv_prepack.cpp:
```diff
@@ -72,6 +74,21 @@ c10::intrusive_ptr<ConvPackedParamsBase<kSpatialDim>> PackedConvWeightCudnn<
+  // TODO: when and if cudnn enables padding in their operators, we can remove padding on our end;
+  // currently, limit padding support to groups=1 (ungrouped conv)
+  // TODO: implement this for groups > 1
+  auto num_output_channels = weight.size(0);
+  if (weight.size(0) % 4 != 0 || weight.size(1) % 4 != 0) {
+    int8_t nslices2pad_out = (4 - weight.size(0) % 4) % 4; // number of slices we need to pad/add for output channels
+    int8_t nslices2pad_in = (4 - weight.size(1) % 4) % 4; // number of slices we need to pad/add for (input channels / groups)
```
we are seeing a lot of weight.size(0) and weight.size(1) here, maybe we can define some variables for them, like output_channels, input_channels to make the code more readable

jerryzh168 replied:
we are seeing a lot of weight.size(0) and weight.size(1) here, maybe we can define some variables for them, like output_channels, input_channels to make the code more readable

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -332,10 +331,26 @@ at::Tensor PackedConvWeightCudnn<kSpatialDim>::apply_impl(
+  // TODO: when and if cudnn enables padding in their operators, we can remove padding on our end;
+  // currently, limit padding support to groups=1 (ungrouped conv)
+  // TODO: implement this for groups > 1; should be straightforward since we're only padding a single dimension
+  if (act.size(1) % 4 != 0) {
+    int8_t num_slices = 4 - act.size(1) % 4; // number of slices we need to pad
```
similar for this one, I think we can assign act.size(1) to a variable

jerryzh168 replied:
similar for this one, I think we can assign act.size(1) to a variable

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/utils.h:
```diff
@@ -168,6 +170,7 @@ struct TORCH_API PackedConvWeightCudnn : public ConvPackedParamsBase<kSpatialDim
   int64_t groups_;
   bool transpose_;
   c10::QScheme q_scheme_;
+  int64_t output_channels_;
```
can we make this more explicit? like unpaded_output_channels_ or original_output_channels_?

jerryzh168 replied:
can we make this more explicit? like unpaded_output_channels_ or original_output_channels_?

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/conv_prepack.cpp:
```diff
@@ -72,6 +74,21 @@ c10::intrusive_ptr<ConvPackedParamsBase<kSpatialDim>> PackedConvWeightCudnn<
+  // TODO: implement this for groups > 1
+  auto num_output_channels = weight.size(0);
+  if (weight.size(0) % 4 != 0 || weight.size(1) % 4 != 0) {
+    int8_t nslices2pad_out = (4 - weight.size(0) % 4) % 4; // number of slices we need to pad/add for output channels
```
can this be simplified to (4 - weight.size(0) % 4)? (removing the extra %4 in the end)

dzdang replied:
not with the way the conditional in the if statement is set up because it's possible for weight.size(0) % 4 to be 0, but I can rewrite it in a better way

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/conv_prepack.cpp:
```diff
@@ -72,6 +74,21 @@ c10::intrusive_ptr<ConvPackedParamsBase<kSpatialDim>> PackedConvWeightCudnn<
+  // TODO: implement this for groups > 1
+  auto num_output_channels = weight.size(0);
+  if (weight.size(0) % 4 != 0 || weight.size(1) % 4 != 0) {
+    int8_t nslices2pad_out = (4 - weight.size(0) % 4) % 4; // number of slices we need to pad/add for output channels
```
can this be simplified to (4 - weight.size(0) % 4)? (removing the extra %4 in the end)

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/conv_prepack.cpp:
```diff
@@ -72,6 +74,21 @@ c10::intrusive_ptr<ConvPackedParamsBase<kSpatialDim>> PackedConvWeightCudnn<
+  if (weight.size(0) % 4 != 0 || weight.size(1) % 4 != 0) {
+    int8_t nslices2pad_out = (4 - weight.size(0) % 4) % 4; // number of slices we need to pad/add for output channels
+    int8_t nslices2pad_in = (4 - weight.size(1) % 4) % 4; // number of slices we need to pad/add for (input channels / groups)
+    weight = at::pad(weight, {0, 0, 0, 0, 0, nslices2pad_in, 0, nslices2pad_out}, "constant", 0);
```
can we have a comment for what this is doing and what does the argument mean?

jerryzh168 replied:
can we have a comment for what this is doing and what does the argument mean?

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -332,10 +331,26 @@ at::Tensor PackedConvWeightCudnn<kSpatialDim>::apply_impl(
+  }
+
+  // need to return sliced tensor if output_channels was padded
+  if (output_channels_ != orig_weight_.size(0)) {
```
should we rename `orig_weight_` to `padded_weight_` now?

dzdang replied:
I don't think that would make sense for all cases because for some cases we're not using padding

jerryzh168 replied:
`maybe_padded_weight_` then?

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -332,10 +331,26 @@ at::Tensor PackedConvWeightCudnn<kSpatialDim>::apply_impl(
+  }
+
+  // need to return sliced tensor if output_channels was padded
+  if (output_channels_ != orig_weight_.size(0)) {
```
should we rename `orig_weight_` to `padded_weight_` now?

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -332,10 +331,26 @@ at::Tensor PackedConvWeightCudnn<kSpatialDim>::apply_impl(
+  if (act.size(1) % 4 != 0) {
+    int8_t num_slices = 4 - act.size(1) % 4; // number of slices we need to pad
+    auto act_padded = at::pad(act, {0, 0, 0, 0, 0, num_slices, 0, 0}, "constant", 0);
+    apply_impl_helper<kReluFused>(
+        quantized_output, act_padded.to(c10::MemoryFormat::ChannelsLast), output_scale);
+  } else {
+    apply_impl_helper<kReluFused>(
+        quantized_output, act.to(c10::MemoryFormat::ChannelsLast), output_scale);
```
we can move this call outside of the `if else block` I think, by having the branch just do padding

jerryzh168 replied:
we can move this call outside of the `if else block` I think, by having the branch just do padding

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -319,7 +319,6 @@ at::Tensor PackedConvWeightCudnn<kSpatialDim>::apply_impl(
-  const int D = kSpatialDim == 3 ? act.size(2) : 1;
   const int H = act.size(kSpatialDim);
   const int W = act.size(kSpatialDim + 1);
   const int M = orig_weight_.size(0); // output channels
```
looks like we already have a variable here for output_channels, maybe we can rename this to make it clearer and use this one

jerryzh168 replied:
looks like we already have a variable here for output_channels, maybe we can rename this to make it clearer and use this one

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -377,7 +392,6 @@ class QConvInt8 final {
       const c10::intrusive_ptr<ConvPackedParamsBase<kSpatialDim>>& packed_weight,
       double output_scale,
       int64_t output_zero_point) {
-    act = act.to(c10::MemoryFormat::ChannelsLast);
```
would be worthwhile to document the expected memoryformat for packed_weight->apply/apply_relu for input I think, also document the memoryformat for the packed weight as well (in packedweight class)

jerryzh168 replied:
would be worthwhile to document the expected memoryformat for packed_weight->apply/apply_relu for input I think, also document the memoryformat for the packed weight as well (in packedweight class)

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/conv_prepack.cpp:
```diff
@@ -81,7 +98,8 @@ c10::intrusive_ptr<ConvPackedParamsBase<kSpatialDim>> PackedConvWeightCudnn<
           transpose,
-          qtype);
+          qtype,
+          num_output_channels);
```
nit: better naming: original or unpadded output channels

jerryzh168 replied:
nit: better naming: original or unpadded output channels

jerryzh168 reviewed (approved):
lg, had some inline naming, formating, code structuring comments

dzdang committed:
Update on "[quant][core][gpu][improvement] Added support for padding quantized cudnn conv2d operator"


Summary:
cudnn v8.4.0 expects input channels for conv2d to be a multiple of 4. If
it is not, we need to explicitly pad it to a multiple of 4 ourselves as
cudnn does not currently support padding intriniscally.
The padding implemented here is limited to groups=1; however, this
should be a straightforward adaption to groups > 1 since we're only
padding a single dimension.

When cudnn enables support for padding, we can remove the padding on our
end.

Test plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

Differential Revision: [D35822988](https://our.internmc.facebook.com/intern/diff/D35822988)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][improvement] Added support for padding quantized cudnn conv2d operator"


Summary:
cudnn v8.4.0 expects input channels for conv2d to be a multiple of 4. If
it is not, we need to explicitly pad it to a multiple of 4 ourselves as
cudnn does not currently support padding intriniscally.
The padding implemented here is limited to groups=1; however, this
should be a straightforward adaption to groups > 1 since we're only
padding a single dimension.

When cudnn enables support for padding, we can remove the padding on our
end.

Test plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

Differential Revision: [D35822988](https://our.internmc.facebook.com/intern/diff/D35822988)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][improvement] Added support for padding quantized cudnn conv2d operator"


Summary:
cudnn v8.4.0 expects input channels for conv2d to be a multiple of 4. If
it is not, we need to explicitly pad it to a multiple of 4 ourselves as
cudnn does not currently support padding intriniscally.
The padding implemented here is limited to groups=1; however, this
should be a straightforward adaption to groups > 1 since we're only
padding a single dimension.

When cudnn enables support for padding, we can remove the padding on our
end.

Test plan:
```
python test/test_quantization.py -k test_qconv2d_cudnn
```

Differential Revision: [D35822988](https://our.internmc.facebook.com/intern/diff/D35822988)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35822988).

dzdang commented:
!mergethis

dzdang commented:
!merge this

dzdang commented:
!merge

dzdang commented:
@pytorchbot merge this please
